### PR TITLE
Create Data Version Control tutorial materials

### PR DESCRIPTION
**Pull request by KristijanIvancic**
**Where to put new files:**

- New files should go into a top-level subfolder, named after the article slug. For example: `my-awesome-article`

**How to merge your changes:** 

1. [Make sure the CI code style tests all pass (+ run the automatic code formatter if necessary).](https://github.com/realpython/materials/blob/master/README.md)
2. Find an RP Team member on Slack and ask them to review & approve your PR.
3. Once the PR has one positive ("approved") review, GitHub lets you merge the PR.
4. 🎉

### CODE REVIEW

File: python-data-version-control/src/evaluate.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+from train import load_images, load_labels, preprocess
+
+
+def main(repo_path):
+    test_csv_path = repo_path / "data/prepared/test.csv"
+    test_df = pd.read_csv(test_csv_path)
+    raw_images = load_images(data_frame=test_df, column_name="filename")
+    labels = load_labels(data_frame=test_df, column_name="label")
+    preprocessed = [preprocess(image) for image in raw_images]
+    test_data = np.concatenate(preprocessed, axis=0)
```
Comment (somacdivad) [inline python-data-version-control/src/evaluate.py:17]: I'm wondering if it would make sense to refactor this out into a function called `load_data()` that returns a tuple of containing the list of labels and the processed images? Something like:

```python
def load_data(data_path):
    df = pd.read_csv(data_path)
    labels = load_labels(data_frame=df, column_name="label")
    raw_images = load_images(data_fram=df, column_name="filename")
    processed_images = [preprocess(image) for image in raw_images]
    data = np.concatenate(processed_images, axis=0)
    return labels, data
```

Then you could just call the function in `main()`.

That might make it easier to describe what's going on when you do your line-by-line breakdown in the article.

Just a thought!

File: python-data-version-control/src/evaluate.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    model_path = repo_path / "model"
+    model = load(model_path / "model.joblib")
```
Comment (somacdivad) [inline python-data-version-control/src/evaluate.py:19]: Splitting this across two lines doesn't feel super necessary to me, especially since you don't use `model_path` a second time:

```suggestion
    model = load(repo_path / "model/model.joblib")
```

File: python-data-version-control/src/evaluate.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    predictions = model.predict(test_data)
+    accuracy = accuracy_score(labels, predictions)
+    metrics = {"accuracy": accuracy}
+    accuracy_path = repo_path / "metrics/accuracy.json"
+    accuracy_path.write_text(json.dumps(metrics))
+
+
+if __name__ == "__main__":
+    repo_path = Path(__file__) / "../.."
+    main(repo_path.resolve())
```
Comment (somacdivad) [inline python-data-version-control/src/evaluate.py:29]: You can get to the repo directory using the `.parent` attribute twice, and then you don't have to call `.resolve()`:

```suggestion
    repo_path = Path(__file__).parent.parent
    main(repo_path)
```

That seems less hacky to me than using `"../.."` and `.resolve()`. I suggest we change that in all of the relevant files.

File: python-data-version-control/src/prepare.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
+from pathlib import Path
+import pandas as pd
+
+folders_to_labels = {"n03445777": "golf ball", "n03888257": "parachute"}
```
Comment (somacdivad) [inline python-data-version-control/src/prepare.py:4]: Let's make this uppercase since we're using it as a constant:

```suggestion
FOLDERS_TO_LABELS = {"n03445777": "golf ball", "n03888257": "parachute"}
```

File: python-data-version-control/src/prepare.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+
+def get_filelist_and_labels(source_path, target_folders):
```
Comment (somacdivad) [inline python-data-version-control/src/prepare.py:7]: It feels a little odd to me that you're taking `target_folders` as an argument here. It seems like you would just check that the folder for an image is in `folders_to_labels` (on line 13) are and not worry about taking any argument.

I guess there might be some architectural reason for doing something like this in a real-world project, but for this example, I think we can simplify it.

```suggestion
def get_filelist_and_labels(source_path):
```

File: python-data-version-control/src/prepare.py [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    images = []
+    labels = []
+    for image_path in source_path.rglob("*/*.JPEG"):
+        filename = image_path.absolute()
+        folder = image_path.parts[-2]
```
Comment (somacdivad) [inline python-data-version-control/src/prepare.py:12]: Instead of subscribing the `.parts` list, it might be more readable to use `.parent.name`:

```suggestion
        folder = image_path.parent.name
```

File: python-data-version-control/src/prepare.py [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+        if folder in target_folders:
+            images.append(filename)
+            label = folders_to_labels[folder]
```
Comment (somacdivad) [inline python-data-version-control/src/prepare.py:15]: This is just to sync the `if` statement with other updates:

```suggestion
        if folder in FOLDERS_TO_LABELS:
            images.append(filename)
            label = FOLDERS_TO_LABELS[folder]
```

File: python-data-version-control/src/prepare.py [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+def main(repo_path):
+    data_path = repo_path / "data"
+    train_path = data_path / "raw/train"
+    test_path = data_path / "raw/val"
+    train_filenames, train_labels = get_filelist_and_labels(
+        train_path, folders_to_labels.keys()
+    )
+    test_filenames, test_labels = get_filelist_and_labels(
+        test_path, folders_to_labels.keys()
+    )
```
Comment (somacdivad) [inline python-data-version-control/src/prepare.py:35]: If you do remove the `target_folders` parameter from `get_filelist_and_labels()`, then you can shorten these lines:

```suggestion
    train_filenames, train_labels = get_filelist_and_labels(train_path)
    test_filenames, test_labels = get_filelist_and_labels(test_path)
```

File: python-data-version-control/src/train.py [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+    return reshaped
+
+
+def main(repo_path):
+    train_csv_path = repo_path / "data/prepared/train.csv"
+    train_df = pd.read_csv(train_csv_path)
+    raw_images = load_images(data_frame=train_df, column_name="filename")
+    labels = load_labels(data_frame=train_df, column_name="label")
+    preprocessed = [preprocess(image) for image in raw_images]
+    train_data = np.concatenate(preprocessed, axis=0)
```
Comment (somacdivad) [inline python-data-version-control/src/train.py:33]: Ah, I just realized that if you refactor this stuff out to the `load_data()` function, then you can just import it in the `evaluate.py` script which will make it even shorter!

File: python-data-version-control/src/evaluate.py [hunk 10]
```diff
@@ -0,0 +1,29 @@
+from joblib import load
+import json
+from pathlib import Path
+import numpy as np
+import pandas as pd
+from sklearn.metrics import accuracy_score
```
[the remark this answers is not in the document]
Comment (KristijanIvancic) [reply python-data-version-control/src/evaluate.py]: Heh, I do this usually (and automatically, with `isort`) but decided to remove the empty spaces because I thought it would be more in line with the style guide. Sometimes it's tough to find the balance between standard coding advice and the RP style guide. 😅

File: python-data-version-control/src/evaluate.py (hunk 1 above)
Comment (KristijanIvancic) [reply python-data-version-control/src/evaluate.py]: Sounds good, done!

File: python-data-version-control/src/evaluate.py (hunk 3 above)
Comment (KristijanIvancic) [reply python-data-version-control/src/evaluate.py]: Yep, that does seem less hacky!

File: python-data-version-control/src/evaluate.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (somacdivad) [reply python-data-version-control/src/evaluate.py]: Yeah, I know it can be a little confusing to know where to deviate from PEP 8 and where to follow it. But that's one of the things TR is for! To help you sort all that out!

**Comment by somacdivad** [comment]
@KristijanIvancic I've approved the PR. No need to merge it or anything. An editor will take care of that before the article goes live!
